aggregator_core/src/datastore.rs

```diff
@@ -1932,10 +1935,38 @@ impl<C: Clock> Transaction<'_, C> {
                     /* round */ &(u16::from(aggregation_job.round()) as i32),
                     /* last_continue_request_hash */
                     &aggregation_job.last_continue_request_hash(),
+                    /* now */ &self.clock.now().as_naive_date_time()?,
                 ],
             )
-            .await?,
-        )
+            .await?;
+        let row_count = rows.len().try_into()?;
+
+        if let Some(row) = rows.into_iter().next() {
+            // We got a row back, meaning we wrote an aggregation job. We check that it wasn't
+            // expired per the task's report_expiry_age, but otherwise we are done.
+            if row.get("is_expired") {
```

Comment:
IIUC this checks whether the interval of time spanned by the reports in this aggregation job is past the report expiry age for the task. Why do we check this _after_ writing an aggregation job instead of not writing the row in the first place?

Reply:
This strategy is used because we need to differentiate between (a) a conflict with an existing row, which due to `ON CONFLICT DO NOTHING` causes nothing to be written and (b) the attempted write being already GC'ed. Case (a) leads to an error (in `check_insert`), while case (b) should silently not write anything (so that writing an aggregation job 1 second before expiry behaves the same as writing an aggregation job 1 second after expiry).

Another option would be to evaluate the expiry before writing anything, but then we'd need to read the task's report expiry age (or have this method take the task/report expiry age) before attempting a write -- this would be additional round-trips for the fast path of "write is not expired".

Reply:
OK, thanks. I don't find this code path particularly objectionable, I just wanted to make sure I understood why we're doing the less-than-screamingly-obvious thing.